### ISSUE TITLE
Pass object back to async function closure

### ISSUE THREAD

**Issue by sdroege** [gtk-rs/gir#578]
Currently we don't pass the object to the async function closure although it's available from the C trampoline.

Having it available there would simplify code quite a bit as otherwise usually tricks with SendCell and the like are needed for getting the object into the closure.

**Comment by EPashkin** [comment]
IMHO this is good idea

**Comment by sdroege** [comment]
CC @antoyo

**Comment by sdroege** [comment]
Might also want to remove the Send bound from async functions, but that needs some deeper analysis first. It *is* valid to first run a main context in one thread and later in another... but nobody does that and it seems like a bad idea anyway, and e.g. in the case of GTK would break everything.

I'll have to check if there are other ways though.

**Comment by tmiasko** [comment]
Even if main context consistently runs in the same thread, the async operation could be scheduled to run there from a different thread, and so closure itself needs to be sent there.

Maybe it would be possible to write a wrapper for context that is not sendable across threads, and should be provided to async operation, which would essentially do push / pop sequence, I wonder what overhead of that would be.

**Comment by sdroege** [comment]
The closure is always called from the thread running the main context, and you schedule it on the main context that is the thread default when calling the async operation. That part should be fine, I'm more worrying about other things, will investigate in the next days :)

**Comment by sdroege** [comment]
Ah right. That was one of the things I wanted to check, but for some reason I thought that a main context can only be thread-default in one thread at a time, and the main loop will already make it thread default.

With that being understood, I think we can close this issue. It would be unsound to pass the object to the closure because you could violate the `Send` constraints (or rather that the object does not implement `Send`, like most GIO things).


However in my work for wrapping futures around all this, we can safely (with some `unsafe` code) do this because the surroundings ensure that such an invalid situation can never happen.

**Comment by tmiasko** [comment]
I must say, in above I was only referring to the Send requirement on the
closure, not to the general idea of passing object to the closure. But, I might
have been inconsistent there.

Either both object should have Send bound and so should closure, or neither
should. Because, by the time we are invoking callback, object was not only
potentially sent to another thread, but also used together with finish method
there. 

That seems quite hard to fit into Rust safety model ... :(

**Comment by sdroege** [comment]
> I must say, in above I was only referring to the Send requirement on the
closure, not to the general idea of passing object to the closure.

Sending a non-Send closure to another thread is fine as long as you don't actually use it from there (i.e. call it). It's just some memory after all.

> Either both object should have Send bound and so should closure, or neither
should. Because, by the time we are invoking callback, object was not only
potentially sent to another thread, but also used together with finish method
there.

The problem here is that all these objects are actually `Send`, just not `Sync`. But something refcounted can't be `Send` unless it is also `Sync` at the same time (otherwise you could clone it and send both clones, the same object, to different threads).

The way how it currently is should be fine though. The object itself might not be `Send` because of the above, but from a Rust point of view you will never have the object in a different thread. Only on the C side it might appear in a different thread (the trampoline calling the closure), but that must be safe from the C API because otherwise the C API would be broken already.

Just be passing the object itself to the Rust closure you would break this.

----
So maybe we should just keep it as-is, and I work around that for the futures implementation where it actually is safe due to being guaranteed that it's all the same thread. It just has slightly more overhead due to converting callbacks into futures, so both approaches are good to have.

**Comment by tmiasko** [comment]
I can see your point about Rust point of view, similar to how things work with
with regards to &self / &mut self thing.

I mostly worried about scenario, where object is not only transferred to a
different thread, but used from two threads at the same time:

Thread A: Dispatches async method on obj.
Thread B: Calls finish on obj.
Thread A: Calls other methods on obj at the same time.

Is above really supposed to work correctly in general, in C API? I would say
not, unless documented otherwise, but that would be possible to write in safe
Rust.  In fact I just looked implementation of a few async methods in GLib /
GIO, the verdict in nonunanimous, some looked safe in this aspect, others
clearly unsafe.

**Comment by sdroege** [comment]
Indeed, it's broken by design of the C API. Simplest example that breaks would be an async read on a socket (which then happens on another thread) directly followed by a synchronous read. Which of the two actually gets the next data is undefined and it's probably even unsafe (calling read simultaneously on the same socket from two different threads). This problem is also not solved by using futures as you can run into it already when being on the specific main context and only calling functions from there.

I think the underlying problem here is, as usual, that everything in GObject is refcounted but GIO/GTK objects are not thread-safe at all, while potentially using other threads nonetheless.

This means that our async functions right now are also potentially unsafe btw, I'm not sure right now how to best fix this, needs some serious thinking :)

**Comment by sdroege** [comment]
2 1/2 options

1) we could split objects into two: one with async, one with sync API. Problem: when getting this from C we don't know which of the two, or worse if C code is using the object in which of the two ways (for Rust code we could store a flag in the object qdata and check that)
1) store a flag in qdata (or even better in GIO itself, but unlikely to happen) to make "pending async operations". We can then assert on that being 0 in all sync operations. Requires understanding which are async operations, which are sync and which are neither and code generator changes.
1) do nothing and just document that you're on your own if you're mixing sync and async operations

Considering all else, I would currently prefer 3). 1) has problems and ugly API, 2) has non-trivial performance impact.

For the problem of getting the object into the async closure, I don't see a way to do that outside of the Futures work I'm currently.

**Comment by antoyo** [comment]
> Simplest example that breaks would be an async read on a socket (which then happens on another thread) directly followed by a synchronous read. Which of the two actually gets the next data is undefined.

Are you sure about that? The doc says:

> During an async request no other sync and async calls are allowed on stream , and will result in G_IO_ERROR_PENDING errors.

**Comment by tmiasko** [comment]
> During an async request no other sync and async calls are allowed on stream , and will result in G_IO_ERROR_PENDING errors.

That wording is really strange, i.e., not allowed followed by description what happens if you try. 
Also, it doesn't say anything about threads, so I wouldn't read this as guarantee of thread safety.
Especially that you can find following in GInputStream docs:

> See the documentation for GIOStream for details of thread safety of streaming APIs.

And then, in GIOStream:

> Operations on GIOStreams cannot be started while another operation on the GIOStream or its substreams is in progress. Specifically, an application can read from the GInputStream and write to the GOutputStream simultaneously (either in separate threads, or as asynchronous operations in the same thread), but an application cannot start any GIOStream operation while there is a GIOStream, GInputStream or GOutputStream operation in progress, and an application can’t start any GInputStream or GOutputStream operation while there is a GIOStream operation in progress.

**Comment by tmiasko** [comment]
There is also option 0, I mentioned earlier. Create non-sendable handle around
context, that can be run only from a single thread, plus a sendable variant
that can only be used to spawn operations that are Send safe.

Require that such handle should be provided to async operation for things that
are not Sendable, so they can do push / pop sequence. (This, unfortunately
presumes that C API respects thread default context, and doesn't just default
to g_main_context_default).

**Comment by tmiasko** [comment]
Time for option 4:

I think, for async methods that are not supposed to be invoked on different
threads we could make it both safe and ergonomic, by dropping Send requirement
on closure and providing object back into it once async method completes.
Thread safety could be established with runtime checks before calling async
function as follows:

* Obtain context with `g_main_context_ref_thread_default()`.
* Check if current thread is the owner of context, if so proceed with
  call, otherwise return error / panic.

This assumes that context will not be run from different threads at different
points in time. It also assumes that async functions will behave like
`g_task_new()` with respect to context used, i.e., use `g_main_context_ref_thread_default()`
to obtain it, which essentially calls `g_main_context_get_thread_default()`
first, and then falls back to `g_main_context_default()`.

**Comment by sdroege** [comment]
That would make sense

> This assumes that context will not be run from different threads at different points in time.

IMHO safe to assume, even if technically possible nobody does that

>  It also assumes that async functions will behave like g_task_new() with respect to context used

Everything else would be a bug in the C function

----

The main problem I see with this is that it requires all code to set the thread default main context, which a lot of code nowadays does not. Especially GTK code usually just uses the global default one and never ever does anything main context related.

Inconveniently also things called from inside the main context don't have a thread default main context set (IIRC)

> Check if current thread is the owner of context, if so proceed with call, otherwise return error / panic.

`g_main_context_is_owner()` you mean? That's IIRC equivalent to having this context as the thread default. `g_main_context_push_thread_default()` acquires the context and becomes its owner, and no other thread can push the same context as thread default until it is popped again (you get an assertion then). Or do you mean to be able to distinguish between "no thread default but we got the global one" and "global one is explicitly thread default and acquired by this thread" (the first being an invalid case that we should panic on)? I guess for that it would work.

In any case, this is basically the "futures" solution I meant. It is ensured that the async operation is always called from a specific main context, and that one is the thread default one so the callback would also go back to the same thread (under the above assumptions).
